libindy_vdr/src/pool/cache/strategy.rs

```diff
@@ -0,0 +1,442 @@
+    expire_after: u128,
+}
+
+impl<K: Eq + Hash + Clone + Send + Sync + 'static, V: Clone + Send + Sync + 'static>
+    CacheStrategyTTL<K, V>
+{
+    /// Create a new cache with the given capacity and expiration time in milliseconds
+    /// If store_type is None, the cache will use an in-memory hashmap and BTreeMap
+    /// cache_time is used as a starting point to generate timestamps if it is None, the cache will use the UNIX_EPOCH as the cache start time
+    pub fn new(
+        capacity: usize,
+        expire_after: u128,
+        store_type: Option<OrderedHashMap<K, u128, V>>,
+        create_time: Option<SystemTime>,
+    ) -> Self {
+        Self {
+            store: match store_type {
+                Some(store) => store,
+                None => OrderedHashMap::new(BTreeMap::new()),
+            },
+            capacity,
+            create_time: match create_time {
+                Some(time) => time,
+                None => SystemTime::UNIX_EPOCH,
+            },
+            expire_after,
+        }
+    }
+}
+
+#[async_trait]
+impl<K: Hash + Eq + Send + Sync + 'static + Clone + Debug, V: Clone + Send + Sync + 'static>
+    CacheStrategy<K, V> for CacheStrategyTTL<K, V>
+{
+    async fn get(&self, key: &K) -> Option<V> {
+        match self.store.get(key) {
+            Some((ts, v)) => {
+                let current_time = SystemTime::now()
+                    .duration_since(self.create_time)
+                    .unwrap()
+                    .as_millis();
+                if current_time < *ts {
+                    Some(v.clone())
+                } else {
+                    None
+                }
+            }
+            None => None,
+        }
+    }
+    async fn remove(&mut self, key: &K) -> Option<V> {
+        self.store.remove(key).map(|(_, v)| v)
+    }
+
+    async fn insert(
+        &mut self,
+        key: K,
+        value: V,
+        config: Option<Vec<CacheStrategyConfig>>,
+    ) -> Option<V> {
+        let current_ts = SystemTime::now()
+            .duration_since(self.create_time)
+            .unwrap()
+            .as_millis();
+
+        // remove expired entries
+        while self.store.len() > 0
+            && self
+                .store
+                .get_first_key_value()
+                .map(|(_, ts, _)| ts.clone() < current_ts)
+                .unwrap_or(false)
+        {
+            self.store.remove_first();
+        }
+
+        // remove the oldest item if the cache is still full
+        if self.store.len() >= self.capacity && self.store.get(&key).is_none() {
+            // remove the oldest item
+            let removal_key = self.store.get_first_key_value().map(|(k, _, _)| k.clone());
+            if let Some(removal_key) = removal_key {
+                self.store.remove(&removal_key);
+            }
+        };
+
+        let exp_offset = match config {
+            Some(opt_vec) => opt_vec
+                .iter()
+                .find(|opt| match opt {
```

Comment:
I think you can use:

```
.find_map(|opt| match opt {
 CacheStrategyConfig::TTL(ttl) => Some(ttl),
 _ => None,
})
```

Reply:
I made it accept a list because the config has to be supplied at the cache level which is strategy agnostic. We don't know what cache strategy is being used under the hood and the user may want it to be at the back of the queue if it's using LRU, or they may want it to expire quickly if it's using TTL. In which case they'd supply a config option for both the LRU and TTL caching strategies